src/category_theory/preadditive/injective.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Kevin Buzzard
+-/
+
+import algebra.homology.exact
+import category_theory.types
+import category_theory.preadditive.projective
+import category_theory.limits.shapes.biproducts
+
+/-!
+# Injective objects and categories with enough injectives
+
+An object `J` is injective iff every morphism into `J` can be obtained by extending a monomorphism.
+-/
+
+noncomputable theory
+
+open category_theory
+open category_theory.limits
+
+universes v u
+
+namespace category_theory
+variables {C : Type u} [category.{v} C]
+
+/--
+An object `J` is injective iff every morphism into `J` can be obtained by extending a monomorphism.
+-/
+class injective (J : C) : Prop :=
+(factors : ∀ {X Y : C} (g : X ⟶ J) (f : X ⟶ Y) [mono f], ∃ h : Y ⟶ J, f ≫ h = g)
+
+section
+/--
+An injective presentation of an object `X` consists of a monomorphism `f : X ⟶ J`
+to some injective object `J`.
+-/
+@[nolint has_inhabited_instance]
+structure injective_presentation (X : C) :=
+(J : C)
+(injective : injective J . tactic.apply_instance)
+(f : X ⟶ J)
+(mono : mono f . tactic.apply_instance)
+
+variables (C)
+
+/-- A category "has enough injectives" if every object has an injective presentation, 
+i.e. if for every object `X` there is an injective object `J` and a monomorphism `X ↪ J`. -/
+class enough_injectives : Prop :=
+(presentation : ∀ (X : C), nonempty (injective_presentation X))
+
+end
+
+namespace injective
+
+/--
+Let `J` be injective and `g` a morphism into `J`, then `g` can be factored through any monomorphism.
+-/
+def factor_of {J X Y : C} [injective J] (g : X ⟶ J) (f : X ⟶ Y) [mono f] : Y ⟶ J :=
+(injective.factors g f).some
```

Comment:
Why did you change the name from `factor_thru`? I'm not opposed, just not sure what the change is meant to indicate.

Reply:
I was thinking that because g equals to f•h, i.e h is a factor of g

Reply:
I prefer `factor_thru`, because (1) you are factoring `g` through `f`, and (2) the word `of` is already overloaded in the mathlib naming convention, so if there is a reasonable alternative, let's pick that.

Reply:
This makes sense, I have changed everything to `factor_thru`